nav2_shim_controller/src/shim_controller.cpp

```diff
@@ -0,0 +1,397 @@
+
+  node->get_parameter(plugin_name_ + ".transform_tolerance", transform_tolerance);
+  node->get_parameter(plugin_name_ + ".max_angular_accel", max_angular_accel_);
+  node->get_parameter(plugin_name_ + ".max_angular_vel", max_angular_vel_);
+  node->get_parameter(plugin_name_ + ".max_linear_vel", max_linear_vel_);
+  desired_linear_vel_ = max_linear_vel_;
+  node->get_parameter(plugin_name_ + ".max_angle_threshold", max_angle_threshold_);
+  node->get_parameter(plugin_name_ + ".lookahead_dist", lookahead_dist_);
+  node->get_parameter(plugin_name_ + ".use_rotate_to_heading", use_rotate_to_heading_);
+  node->get_parameter(plugin_name_ + ".use_rotate_to_path", use_rotate_to_path_);
+  node->get_parameter(plugin_name_ + ".use_dynamic_threshold", use_dynamic_threshold_);
+  node->get_parameter("controller_frequency", control_frequency);
+
+  transform_tolerance_ = tf2::durationFromSec(transform_tolerance);
+  control_duration_ = 1.0 / control_frequency;
+
+  if (node->get_parameter(plugin_name_ + ".default_plugin", default_plugin_name)) {
+    std::string default_plugin_type = nav2_util::get_plugin_type_param(node, default_plugin_name);
+    default_plugin = plugin_loader_.createSharedInstance(default_plugin_type);
+    RCLCPP_INFO(
+      logger_, "Initializing default plugin \"%s\" with type \"%s\"",
+      default_plugin_name.c_str(), default_plugin_type.c_str());
+
+    try {
+      default_plugin->configure(node, default_plugin_name, tf_, costmap_ros_);
+    } catch (const std::exception & e) {
+      RCLCPP_ERROR(logger_, "Couldn't initialize default plugin! %s", e.what());
+      throw nav2_core::PlannerException("Unable to load default plugin!");
+    }
+    RCLCPP_INFO(logger_, "Initialized default plugin \"%s\"", default_plugin_name.c_str());
+  } else {
+    RCLCPP_ERROR(
+      node->get_logger(), "Can not get 'default_plugin' param value for %s", plugin_name_.c_str());
+  }
+}
+
+void ShimController::cleanup()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Cleaning up controller: %s of type"
+    " shim_controller::ShimController",
+    plugin_name_.c_str());
+
+  if (default_plugin) {default_plugin->cleanup();}
+}
+
+void ShimController::activate()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Activating controller: %s of type "
+    "shim_controller::ShimController",
+    plugin_name_.c_str());
+  if (default_plugin) {default_plugin->activate();}
+}
+
+void ShimController::deactivate()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Deactivating controller: %s of type "
+    "shim_controller::ShimController",
+    plugin_name_.c_str());
+
+  if (default_plugin) {default_plugin->deactivate();}
+}
+
+geometry_msgs::msg::TwistStamped ShimController::computeVelocityCommands(
+  const geometry_msgs::msg::PoseStamped & pose,
+  const geometry_msgs::msg::Twist & speed,
+  nav2_core::GoalChecker * goal_checker)
+{
+  goal_checker_ = goal_checker;
+  return calcCmdVel(pose, speed);
+}
+
+geometry_msgs::msg::TwistStamped ShimController::calcCmdVel(
```

Comment:
Put this function into `computeVelocityCommands`, there's no real reason to separate this out.

Reply:
Then goal_checker doesn't need to be a class member anymore